### PR TITLE
Add QBFT Propose and RoundChange messages

### PR DESCRIPTION
**Pull request by rain-on**
This adds the proposal and roundchange messages required for QBFT.

This does _not_ wire in the ability to transmit/receive these messages, it just defines the message structure.

- [ X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/messagewrappers/RoundChangeTest.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      new BftExtraData(
+          Bytes32.ZERO, Collections.emptyList(), Optional.empty(), 1, Collections.emptyList());
+
+  private static final Block BLOCK =
+      new Block(
+          new BlockHeaderTestFixture().extraData(extraData.encode()).buildHeader(),
+          new BlockBody(Collections.emptyList(), Collections.emptyList()));
+
+  @Test
+  public void canRoundTripARoundChangeMessage() {
```
Comment (jframe) [inline consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/messagewrappers/RoundChangeTest.java:54]: Can a test case be added for the empty prepared round and prepared value case

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  /**
+   * Constructor that derives the sequence and round information from an RLP encoded message
+   *
+   * @param in The RLP body of the message to check
+   * @return A derived sequence and round number
+   */
+  public static PreparedRoundMetadata readFrom(final RLPInput in) {
+    in.enterList();
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java:49]: Is extra level of nesting needed? I think originally we were proposing a flatter structure for the preparedValue and preparedRound.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  public int getPreparedRound() {
+    return preparedRound;
+  }
+
+  /**
+   * Constructor that derives the sequence and round information from an RLP encoded message
+   *
+   * @param in The RLP body of the message to check
+   * @return A derived sequence and round number
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java:46]: This doc seems incorrect, we aren't returning a (height, round)

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import org.hyperledger.besu.ethereum.rlp.RLPInput;
+import org.hyperledger.besu.ethereum.rlp.RLPOutput;
+
+import java.util.Objects;
+
+import com.google.common.base.MoreObjects;
+
+public class PreparedRoundMetadata {
+  private final Hash preparedBlockHash;
+  private final int preparedRound;
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java:27]: This should be a big integer

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/RoundChangeCertificate.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.consensus.qbft.payload;
+
+import org.hyperledger.besu.consensus.common.bft.payload.SignedData;
+
+import java.util.Collection;
+import java.util.StringJoiner;
+
+public class RoundChangeCertificate {
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/RoundChangeCertificate.java:22]: Didn't think the RoundChange certificate was needed for QBFT?

File: consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/messagewrappers/RoundChangeTest.java (hunk 1 above)
Comment (rain-on) [reply consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/messagewrappers/RoundChangeTest.java:54]: done

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java (hunk 3 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java:46]: we can go flatter - done.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java (hunk 2 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java:49]: removed

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java (hunk 4 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java:27]: round as a big integer is _such_ an unrequired change - i.e. it implies we go into roundchanges for millenia.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/RoundChangeCertificate.java (hunk 5 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/RoundChangeCertificate.java:22]: it makes the creation of messages somewhat simplier, as it collates the various data required to construct a proposal message.

File: consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/messagewrappers/RoundChangeTest.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+    final RoundChangePayload payload =
+        new RoundChangePayload(new ConsensusRoundIdentifier(1, 1), Optional.empty());
+
+    final SignedData<RoundChangePayload> signedRoundChangePayload =
+        new SignedData<>(payload, addr, nodeKey.sign(MessageFactory.hashForSignature(payload)));
+
+    final PreparePayload preparePayload =
+        new PreparePayload(new ConsensusRoundIdentifier(1, 0), BLOCK.getHash());
+    final SignedData<PreparePayload> signedPreparePayload =
```
Comment (jframe) [inline consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/messagewrappers/RoundChangeTest.java:103]: nit: this variable is unused

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/MessageFactory.java [hunk 7]
```diff
@@ -34,19 +42,73 @@ public MessageFactory(final NodeKey nodeKey) {
...
+              .getPayload()
+              .getRoundIdentifier()
+              .getRoundNumber();
+
+      payload =
+          new RoundChangePayload(
+              roundIdentifier,
+              Optional.of(
+                  new PreparedRoundMetadata(
+                      preparedRoundData.get().preparedBlock.getHash(), BigInteger.valueOf(round))));
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/MessageFactory.java:99]: The ConsensusRoundIdentifier needs to use BigInteger as well. But perhaps another PR for that change?

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/RoundChangeCertificate.java (hunk 5 above)
Comment (jframe) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/RoundChangeCertificate.java:22]: Think a struct makes sense, but calling it a RoundChangeCertificate is a bit confusing as that is for a different purpose or at least that's what I remember. RoundChangePayMetadata?

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/MessageFactory.java (hunk 7 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/MessageFactory.java:99]: I think another PR would be good - need to think about what that means for IBFT2 ... i.e. can we reuse this struct, or not?

File: consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/messagewrappers/RoundChangeTest.java (hunk 6 above)
Comment (rain-on) [reply consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/messagewrappers/RoundChangeTest.java:103]: repaired.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messagewrappers/Proposal.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+  public List<SignedData<PreparePayload>> getPrepares() {
+    return prepares;
+  }
+
+  public Block getBlock() {
+    return getPayload().getProposedBlock();
+  }
+
+  public Optional<RoundChangeMetadata> getRoundChangeCertificate() {
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messagewrappers/Proposal.java:60]: nit: rename the method to getRoundChangeMetadata to match the type

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/MessageFactory.java [hunk 9]
```diff
@@ -34,19 +41,73 @@ public MessageFactory(final NodeKey nodeKey) {
     this.nodeKey = nodeKey;
   }
 
+  public Proposal createProposal(
+      final ConsensusRoundIdentifier roundIdentifier,
+      final Block block,
+      final Optional<RoundChangeMetadata> roundChangeCertificate) {
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/MessageFactory.java:47]: nit: rename var to roundChangeMetadata

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messagewrappers/Proposal.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.tuweni.bytes.Bytes;
+
+public class Proposal extends BftMessage<ProposalPayload> {
+
+  private final List<SignedData<RoundChangePayload>> roundChanges;
+  private final List<SignedData<PreparePayload>> prepares;
+
+  public Proposal(
+      final SignedData<ProposalPayload> payload,
+      final List<SignedData<RoundChangePayload>> roundChanges,
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messagewrappers/Proposal.java:41]: Is the order important? Would make the code in the MessageFactory simpler if we had List/Collection types consistent between the Proposal and the RoundChangeMetadata

**Review by jframe — APPROVED** [review]

**Pull request merged**